### ISSUE TITLE
Slow `FermionOperator2nd` manipulation

### ISSUE THREAD

**Issue by PhilipVinc** [netket/netket#1576]
@gcarleo recently noticed that algebraic manipulations of `FermionOperator2nd` is very slow compared to `PauliStrings` and sometimes it is impossible to write some (albeit complex) hamiltonians for as few as 16 orbitals.

This is just about the symbolic manipulations, and not about constructing (`_setup`) the representation that we use for computing connected elements.

benchmarking using `py-spy` I suspect that there are two issues:
 - The data-structure used to store the terms is a dictionary, where keys are the 'terms' and the values are the coefficients in the sum. This is different from `PauliStrings` which uses two lists for terms and coefficients, which makes manipulations somewhat cheaper.
 - `PauliStrings` prunes zeros and repeated entries less often than `FermionOperator2nd`, saving time.

I think fixing this issue should not be very hard, and as a first-order change we could just change the data structure used.

**Comment by jwnys** [comment]
iirc I just numba jitted the rules and logic for PauliStrings manipulations, and not for FermionOperator2nd.

**Comment by jwnys** [comment]
Btw, how about doing fermi hubbard just in a way like this (this is the jax version, the numba version is easier)
```python
import jax
import jax.numpy as jnp
from functools import partial
import numpy as np
from collections import defaultdict

from flax import struct
import warnings
from jax.tree_util import register_pytree_node_class

from netket.utils.types import PyTree
from netket.operator import DiscreteJaxOperator
from netket.experimental.hilbert import SpinOrbitalFermions


@register_pytree_node_class
class FermiHubbardJax(DiscreteJaxOperator):
    def __init__(self, hilbert, graph, t=1, U=0, V=0, Vext=0):
        assert isinstance(hilbert, SpinOrbitalFermions)
        self._t = t
        self._U = U # opposite spin interaction, on site
        self._V = V # equal spin interactions, nearest neighbor
        self._Vext = Vext

        self._graph = graph
        self._max_conn_size = None
        self._n_fermions_per_spin = None
        self._n_fermions = None
        self._n_spin_states = None
        self._edges_list = None
        self._edges = None
        self._initialized = False
        self._dtype = jnp.float64 # keep fixed, only then hermitian
        super().__init__(hilbert)
        
    def tree_flatten(self):
        self._setup()
        array_data = (
            self._t,
            self._U,
            self._V,
            self._Vext,
            self._edges_list, # rather a hashable array???
            self._edges,
        )
        struct_data = {
            'hilbert': self.hilbert,
            'max_conn_size' : self._max_conn_size,
            'n_fermions' : self._n_fermions,
            'n_fermions_per_spin' : self._n_fermions_per_spin,
            'n_spin_states' : self._n_spin_states,
        }
        return array_data, struct_data

    @classmethod
    def tree_unflatten(cls, struct_data, array_data):
        t, U, V, Vext, edges_list, edges = array_data
        hilbert = struct_data['hilbert']
        graph = None
        obj = cls(hilbert, graph, t=t, U=U, V=V, Vext=Vext)
        obj._max_conn_size = struct_data["max_conn_size"]
        obj._n_fermions_per_spin = struct_data["n_fermions_per_spin"]
        obj._n_fermions = struct_data["n_fermions"]
        obj._n_spin_states = struct_data["n_spin_states"]
        obj._edges_list = edges_list
        obj._edges = edges
        obj._initialized = True
        return obj


    def _setup(self, force=False):
        if force or not self._initialized:
            n_nodes = self.hilbert.n_orbitals
            n_modes = self.hilbert.size
            assert n_nodes == self._graph.n_nodes
            n_fermions_per_spin = self.hilbert.n_fermions
            assert n_fermions_per_spin is not None
            if not hasattr(n_fermions_per_spin, '__iter__'):
                n_fermions_per_spin = (n_fermions_per_spin, )
            self._n_fermions_per_spin = tuple(list(map(int, n_fermions_per_spin)))
            self._n_fermions = sum(n_fermions_per_spin)
            self._n_spin_states = len(n_fermions_per_spin)
            self._has_spin = self.hilbert.spin is not None

            # off diagonal part
            self._t = jnp.asarray(self._t)
            if self._t.size == 1:
                self._t = jnp.full((n_nodes, n_nodes), self._t.reshape(-1))
            assert self._t.shape == (n_nodes, n_nodes), f"t={self._t.shape}"
            
            # at half filling or lower, this is the maximum amount of moves per spin
            self._edges = np.array(list(self._graph.edges()))
            _, Ks = np.unique(self._edges, return_counts=True) # get the degree
            assert np.all(Ks[0] == Ks) # we assume constant degree
            Kmax = np.max(Ks) # maximum degree of the lattice
            edge_dict = defaultdict(list)
            for u, v in self._graph.edges():
                edge_dict[u].append(v)
                edge_dict[v].append(u)
            self._edges_list = np.array([edge_dict[u] for u in range(n_nodes)])
            # for faster jw strings, we rely on it
            self._edges_list = np.sort(self._edges_list, axis=-1)
            assert self._edges_list.shape[-1] == Kmax, f"got {self._edges_list.shape} vs {Kmax}"
            
            # diagonal part
            if not self._has_spin:
                if self._U is not None and not np.isclose(self._U, 0).all():
                    warnings.warn("cannot have nonzero U if the hilbert space is spinless, setting U=0")
                self._U = None
            else:
                self._U = jnp.asarray(self._U).ravel()
                if self._U.size == 1:
                    self._U = jnp.repeat(self._U, n_nodes)
                assert self._U.size == self.hilbert.n_orbitals
            
            self._V = jnp.asarray(self._V)
            if self._V.size == 1:
                self._V = jnp.full((n_nodes, n_nodes), self._V.reshape(-1))
            assert self._V.shape == (n_nodes, n_nodes)
            
            # if callable(self._Vext):
            #     self._Vext = [self._Vext(u) for u in range(n_modes)]
            self._Vext = np.array(self._Vext).reshape(-1)
            if self._Vext.size == 1:
                self._Vext = np.repeat(self._Vext, n_modes)
            assert self._Vext.size == n_modes
            
            # assume spin conserving hopping
            max_conn_sizes = [nf*Kmax for nf in self._n_fermions_per_spin]
            self._max_conn_size = sum(max_conn_sizes) + 1 # incl diagonal
            
        self._initialized = True
        
    @property
    def max_conn_size(self):
        self._setup()
        return self._max_conn_size

    @property
    def dtype(self):
        return self._dtype

    @property
    def is_hermitian(self):
        return True

    def operator_string(self):
        return f"FermiHubbardJax(hilbert={self.hilbert})"
        
    def get_conn_padded(self, x):
        self._setup()
        assert x.ndim == 2
        if isinstance(x, np.ndarray):
            x = jnp.asarray(x)       

        # move things per spin
        x_per_spin = jnp.split(x, self._n_spin_states, axis=-1)
        xp, mels = list(zip(*[move_particles(xs, self._t, nfs, self._edges_list) for xs, nfs in zip(x_per_spin, self._n_fermions_per_spin)]))
        if len(xp) == 2:
            # reconcatenate
            xp = [
                jnp.concatenate(jnp.broadcast_arrays(xp[0], x_per_spin[1][:,None,:]), axis=-1),
                jnp.concatenate(jnp.broadcast_arrays(x_per_spin[0][:,None,:], xp[1]), axis=-1)
            ]
        xp = jnp.concatenate(xp, axis=-2)
        mels = jnp.concatenate(mels, axis=-1) 
        mels = -mels# -t hopping !!!
        
        # diagonal part       
        # V (always, for every spin, vmap instead?)
        mels_diag = sum(_compute_V(xi, self._V, self._edges) for xi in x_per_spin)
        
        # U
        if self._U is not None:
            assert len(x_per_spin) == self._n_spin_states
            mels_diag += _compute_U(self._U, *x_per_spin)
            
        # Vext
        mels_diag += _compute_Vext(self._Vext, x)
        
        # diag first (!)
        xp = jnp.concatenate([x[:,None,:], xp], axis=-2)
        mels = jnp.concatenate([mels_diag[:,None], mels], axis=-1)
        return xp, mels

@partial(jax.jit, static_argnames=("nf",))
def move_particles(x, h, nf, edges_list): # x must be for a given spin only
    assert x.ndim == 2
    Kmax = edges_list.shape[-1]
    
    def move_particles_sample(x):
        occ_idxs = jnp.where(x, size=nf)[0]
        # move each particle
        _hopper = partial(xp_and_mel_offdiag_sample, h, Kmax)
        xp, mels = jax.vmap(_hopper , in_axes=(None, 0, 0), out_axes=0)(x, occ_idxs, edges_list[occ_idxs,:])
        xp = xp.reshape(-1, xp.shape[-1])
        mels = mels.reshape(-1)
        assert xp.shape[-2] == mels.shape[0]
        return xp, mels
    
    return jax.vmap(move_particles_sample)(x)

def xp_and_mel_offdiag_sample(t, Kmax, x, idx, conn_idxs): # make xp even if not feasible
    assert conn_idxs.size == Kmax
    allowed = x[idx] != x[conn_idxs]
    xp = x.at[idx].set(0)
    xp = jnp.repeat(xp[None, :], Kmax, axis=0)
    xp = xp.at[jnp.arange(Kmax), conn_idxs].set(1)
    jw = jax.vmap(jw_sample, in_axes=(None, None, 0), out_axes=0)(x, idx, conn_idxs)
    mels = t[idx, conn_idxs]*jw
    mels *= allowed
    xp = xp*allowed[:,None] + x[None,:]*(~allowed[:,None]) # to keep fermion number
    return xp, mels


@jax.jit
def jw_sample(x, create_idx, destroy_idx):
    m = jnp.arange(x.size)
    mask_sites_between_smaller = (m < create_idx) & (m >= destroy_idx)
    mask_sites_between_larger = (m > create_idx[..., None]) & (m <= destroy_idx)
    mask_sites_between = mask_sites_between_smaller | mask_sites_between_larger
    sgn = (-1)**jnp.sum(mask_sites_between*x) # 1 and in between --> -1
    return sgn

@jax.jit
def _compute_V(x, V, edges):
    ni = x[:, edges[:,0]]
    nj = x[:, edges[:,1]]
    nij = ni*nj
    Vij = V[edges[:,0], edges[:,1]]
    return nij.dot(Vij)

@jax.jit
def _compute_U(U, x_up, x_down):
    nij = x_up*x_down
    return nij.dot(U)

@jax.jit
def _compute_Vext(Vext, x):
    return x.dot(Vext)        
```

**Comment by PhilipVinc** [comment]
Closed (mostly) by #1639
